src/common/eigf.h

```diff
@@ -0,0 +1,302 @@
+dt_omp_firstprivate(guide, mask, Ndim) \
+  schedule(simd:static) aligned(guide, mask:64) \
+  reduction(max:maxg, maxm)\
+  reduction(min:ming, minm)
+#endif
+  for(size_t k = 0; k < Ndim; k++)
+  {
+    const float pixelg = guide[k];
+    const float pixelm = mask[k];
+    if(pixelg < ming) ming = pixelg;
+    if(pixelg > maxg) maxg = pixelg;
+    if(pixelm < minm) minm = pixelm;
+    if(pixelm > maxm) maxm = pixelm;
+  }
+
+  dt_gaussian_t *g = dt_gaussian_init(width, height, 1, &maxg, &ming, sigma, 0);
+  if(!g) return;
+  dt_gaussian_blur(g, guide, blurred_guide);
+  dt_gaussian_free(g);
+
+  g = dt_gaussian_init(width, height, 1, &maxm, &minm, sigma, 0);
+  if(!g) return;
+  dt_gaussian_blur(g, mask, blurred_mask);
+  dt_gaussian_free(g);
+
+  float mingg = 10000000.0f;
+  float maxgg = 0.0f;
+  float mingm = 10000000.0f;
+  float maxgm = 0.0f;
+#ifdef _OPENMP
+#pragma omp parallel for simd default(none) \
+  dt_omp_firstprivate(guide, mask, blurred_guide, blurred_mask, guide_x_guide, guide_x_mask, Ndim) \
+  schedule(simd:static) aligned(guide, mask, blurred_guide, blurred_mask, guide_x_guide, guide_x_mask:64) \
+  reduction(max:maxgg, maxgm)\
+  reduction(min:mingg, mingm)
+#endif
+  for(size_t k = 0; k < Ndim; k++)
+  {
+    const float deviation = guide[k] - blurred_guide[k];
+    const float squared_deviation = deviation * deviation;
+    guide_x_guide[k] = squared_deviation;
+    if(squared_deviation < mingg) mingg = squared_deviation;
+    if(squared_deviation > maxgg) maxgg = squared_deviation;
+    const float cov = (guide[k] - blurred_guide[k]) * (mask[k] - blurred_mask[k]);
+    guide_x_mask[k] = cov;
+    if(cov < mingm) mingm = cov;
+    if(cov > maxgm) maxgm = cov;
+  }
+
+  g = dt_gaussian_init(width, height, 1, &maxgg, &mingg, sigma, 0);
+  if(!g) return;
+  dt_gaussian_blur(g, guide_x_guide, guide_variance);
+  dt_gaussian_free(g);
+
+  g = dt_gaussian_init(width, height, 1, &maxgm, &mingm, sigma, 0);
+  if(!g) return;
+  dt_gaussian_blur(g, guide_x_mask, guide_mask_covariance);
+  dt_gaussian_free(g);
+
+  float mina = 10000000.0f;
+  float minb = 10000000.0f;
+  float minw = 10000000.0f;
+  float maxa = 0.0f;
+  float maxb = 0.0f;
+  float maxw = 0.0f;
+#ifdef _OPENMP
+#pragma omp parallel for simd default(none) \
+  dt_omp_firstprivate(guide, mask, blurred_guide, guide_variance, blurred_mask, guide_mask_covariance, a, b, weights, Ndim, feathering) \
+  schedule(simd:static) aligned(guide, mask, blurred_guide, guide_variance, blurred_mask, guide_mask_covariance, a, b, weights:64) \
+  reduction(max:maxa, maxb, maxw)\
+  reduction(min:mina, minb, minw)
+#endif
+  for(size_t k = 0; k < Ndim; k++)
+  {
+    const float pixelg = fmaxf(guide[k], 1E-6);
+    const float pixelm = fmaxf(mask[k], 1E-6);
+    const float normalized_var_guide = guide_variance[k] / (pixelg * pixelg);
+    const float normalized_covar = guide_mask_covariance[k] / (pixelg * pixelm);
+    // empirical weight
+    float w = 1.f / pixelg;
```

Comment:
I wonder if `w = 1.f / (blurred_mask[k] * blurred_guide[k])` would not work better.

Reply:
here is a comparison
using `w = 1.f / (blurred_mask[k] * blurred_guide[k])`
![Capture d’écran du 2020-10-05 16-18-29](https://user-images.githubusercontent.com/34063828/95091335-9e3ece80-0726-11eb-9fc3-5bc94e4cb0a0.png)
using `w = 1.f / (mask[k] * guide[k])`
![Capture d’écran du 2020-10-05 16-19-25](https://user-images.githubusercontent.com/34063828/95091338-9ed76500-0726-11eb-9368-a14f91677347.png)

The problem with `w = 1.f / (blurred_mask[k] * blurred_guide[k])` is that `blurred_guide` (and `blurred_mask`) is a very bad approximation of guide in the dark area which are close to bright areas

Reply:
ok, the example is self-explanatory. Thanks.